### PR TITLE
Cache friendly reduceop

### PR DESCRIPTION
**Pull request by SzymonOzog**
Currently reduce op starts with acc goes with loded data and ends with acc, It's much more cache friendly if we do local data first and then end with acc, PTX speed vs cuda down to 0.94 with this + https://github.com/tinygrad/tinygrad/pull/4304. Also fixes test_failure_27

### CODE REVIEW

**Comment by chenyuxyz** [comment]
looks like RHIP fails but real HSA works, @Qazalin fyi
image

**Comment by Qazalin** [comment]
ack - fixing

**Comment by SzymonOzog** [comment]
> Can you paste a diff so I can understand what this is doing?

sure thing, it transforms
```
        fma.rn.f32      %alu_f32_60, %val_f32_12, %val_f32_19, %acc_f32_15; # load acc into L1 cache 
        fma.rn.f32      %alu_f32_61, %val_f32_13, %val_f32_23, %alu_f32_60;
        fma.rn.f32      %alu_f32_62, %val_f32_14, %val_f32_27, %alu_f32_61;
        fma.rn.f32      %alu_f32_63, %val_f32_15, %val_f32_31, %alu_f32_62;
        mov.b32         %acc_f32_15, %alu_f32_63; #load acc into L1 cache again
```
into 
```
        mul.f32         %alu_f32_60, %val_f32_13, %val_f32_23;
        fma.rn.f32      %alu_f32_61, %val_f32_12, %val_f32_19, %alu_f32_60;
        fma.rn.f32      %alu_f32_62, %val_f32_14, %val_f32_27, %alu_f32_61;
        fma.rn.f32      %alu_f32_63, %val_f32_15, %val_f32_31, %alu_f32_62;
        add.f32         %alu_f32_64, %alu_f32_3, %acc_f32_0; #  load acc into L1 cache 
        mov.b32         %acc_f32_0, %alu_f32_64; # acc is already in L1 cache
```
the comments describe the problem. In this example even when kernel 1 managed to use `fma` the memory layout is so bad that it makes it terribly slow. Comparison:
kernel1: `RATIO 2.43016177651771 us 73099.26`
kernel2: `RATIO 0.8328376209573036 us 27662.33`

File: tinygrad/codegen/linearizer.py [hunk 1]
```diff
@@ -444,11 +444,15 @@ def ast_parse(self, x:LazyOp, acc: List[UOp], offs:Optional[List[int]], loaded_b
       ret: List[UOp] = []
       input_acc = acc[:]
       for val, off in zip(zip(*values), cast(List[int], offs)):
-        acc[off] = self.uops.add(UOps.ALU, acc[off].dtype, vin=val+(acc[off],), arg=ops[cast(ReduceOps, x.op)], insert_before=insert_before)
+        assert(len(val)==1)
+        acc[off] = val[0] if acc[off] in input_acc else self.uops.add(UOps.ALU, acc[off].dtype, vin=val+(acc[off],), \
```
[the remark this answers is not in the document]
Comment (SzymonOzog) [reply tinygrad/codegen/linearizer.py]: simplified and added explanation

**Comment by chaosagent** [comment]
aren't accs in register? seems like this allows better register coloring instead of L1 cache locality.

**Comment by SzymonOzog** [comment]
> aren't accs in register? seems like this should only allow using less registers instead of L1 cache locality (but i don't see a reason why this should decrease registers either. maybe the acc regs are being spilled?)

I might be wrong on this one but profiling with nsight reported 50% more cache misses on the first kernel. I'm on my phone so I can't post a screenshot but they had a warning inside that excesive registers are moved to local memory

**Comment by geohot** [comment]
Yea, I don't see why this has anything to do with L1 cache.

**Comment by SzymonOzog** [comment]
Okay I've investigated that further and it seems that my initial assumptions were wrong. There are no register spills happening so no local memory is being used for acc. There is a better cache utilization after this change(image below is a comparison from nsight-compute) 
image
But the reason seems to be going from PTX to SASS. [ptxas optimizes instruction schedules](https://forums.developer.nvidia.com/t/ptx-instructions-are-reordered/197973) and it manages to perform better when there is more locality in accessing registers
For a comparison this is a generated SASS before this change
https://pastebin.com/raw/9Mtj3VCx
and after:
https://pastebin.com/raw/RjnLxh94
The important part is line 67 this is where the throttling occurs, right now it groups too much loads together. I've looked at CUDA output and it also generates the same register access pattern as this pr

**Pull request closed without merging**